README.md
# Inline Column Configuration plugin for MantisBT

Copyright (c) 2011  Robert Munteanu - robert@lmn.ro  
Copyright (c) 2015  Carlos Proensa

Released under the [GNU GPL v2 licence](http://opensource.org/licenses/GPL-2.0)

For further details, please refer to the [plugin's page](https://www.mantisbt.org/wiki/doku.php/mantisbt:inlinecolumnconfiguration)
in the MantisBT wiki.

See [Changelog](https://github.com/mantisbt-plugins/InlineColumnConfiguration/blob/master/CHANGELOG.md).


## Description

The InlineColumnConfiguration allows you to configure the column preferences 
inline, in the view bugs page, instead of going through the _Manage Columns_
page in user preferences.


## Requirements

The plugin requires 

  * [MantisBT](http://mantisbt.org/) version 2.2 or higher
  * [jQueryUI plugin](https://github.com/mantisbt-plugins/jQuery-UI) 
    version 1.8 or higher

If you are running MantisBT 1.x, please use 
[Version 1.2.0](https://github.com/mantisbt-plugins/InlineColumnConfiguration/tree/v1.2.0).


## Installation

1. Download the plugin's [latest release](https://github.com/mantisbt-plugins/InlineColumnConfiguration/releases/latest) 
   or clone the [source code](https://github.com/mantisbt-plugins/InlineColumnConfiguration).
2. Copy or link the plugin (the `InlineColumnConfiguration/` directory) into your Mantis
   installation's `plugins/` directory.
3. While logged into your Mantis installation as an administrator, go to
   *Manage -> Manage Plugins*.
4. In the *Available Plugins* list, you'll find the *InlineColumnConfiguration* 
   plugin; click the **Install** link.


## Usage

The plugin adds a *Configure columns* menu item to the View issues page.

When clicked, a configuration pop-up dialog opens, allowing users to select
which fields to display for _View_, _Print_, _CSV_ and _Excel_ columns, as well
as drag and drop them to control the order in which they are displayed.


## Support

The latest source code can found on
[Github](https://github.com/mantisbt-plugins/InlineColumnConfiguration).

We encourage you to submit Bug reports and enhancements requests on the
[MantisBT issues tracker, under the *Plugin - InlineColumnConfiguration* project](https://www.mantisbt.org/bugs/view_all_bug_page.php?project_id=14).
If you would like to propose a patch, do not hesitate to submit a new
[Pull Request](https://github.com/mantisbt-plugins/InlineColumnConfiguration/compare/).


CHANGELOG.md
# Inline Columns Configuration Plugin Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/)
specification.

--------------------------------------------------------------------------------

## [1.2.0] - 2018-09-06

### Added

- Use AJAX to load and save form
- Spanish translation

### Fixed

- Prevent protected users from changing column preferences
- Always use the same Project ID when generating the form
- Incorrect plugin initialization if user subsystem is not ready
- Order custom fields by name


## [1.1.0] - 2015-10-24

### Added

- MantisBT 1.3 compatibility [#20222](https://mantisbt.org/bugs/view.php?id=20222)

### Changed

- Add project name to pop-up window title [#14513](https://mantisbt.org/bugs/view.php?id=14513)

### Fixed

- Display disabled link for ALL_PROJECTS [#14511](https://mantisbt.org/bugs/view.php?id=14511)


## [1.0.1] - 2011-09-11

### Fixed

- Plugin does not work user based [#14096](https://mantisbt.org/bugs/view.php?id=14096)
- custom fields with capitalized names appear twice [#14777](https://mantisbt.org/bugs/view.php?id=14777)

 
## [1.0.0] - 2011-09-11

### Added

- Initial release


[Unreleased]: https://github.com/mantisbt-plugins/InlineColumnConfiguration/compare/v1.2.0...HEAD

[1.2.0]: https://github.com/mantisbt-plugins/InlineColumnConfiguration/compare/v1.1...v1.2.0
[1.1.0]: https://github.com/mantisbt-plugins/InlineColumnConfiguration/compare/v1.0.1...v1.1
[1.0.1]: https://github.com/mantisbt-plugins/InlineColumnConfiguration/compare/v1.0...v1.0.1
[1.0.0]: https://github.com/mantisbt-plugins/InlineColumnConfiguration/commits/v1.0



pages/ajax_form.php
<?php

function print_columns_inputs_by_target ( $p_target ) {
	global $g_project_override;
	$t_project_id = $g_project_override;

	# Instead of columns_get_all(), get separately, to have the column groups ordered
	# $t_all_columns = columns_get_all( $t_project_id );

	$t_columns_std = columns_get_standard();

	$t_columns_plugin = array_keys( columns_get_plugin_columns() );
	sort( $t_columns_plugin );

	$t_columns_cf = array();
	$t_related_custom_field_ids = custom_field_get_linked_ids( $t_project_id );
	foreach( $t_related_custom_field_ids as $t_id ) {
		if( !custom_field_has_read_access_by_project_id( $t_id, $t_project_id ) ) {
			continue;
		}
		$t_def = custom_field_get_definition( $t_id );
		$t_columns_cf[] = 'custom_' . $t_def['name'];
	}
	sort( $t_columns_cf  );

	$t_all_columns = array_merge( $t_columns_std, $t_columns_plugin, $t_columns_cf );

	$t_view_columns = helper_get_columns_to_view( $p_target, /* $p_viewable_only */ false);
	$t_deselected_columns = array_diff( $t_all_columns, $t_view_columns );

	$f_token = gpc_get('ajax_form_token');
	?>
	<form method="post" action="<?php echo plugin_page( 'ajax_form_update' ) ?>">
		<div class="columns_form_actions">
			<div class="action_div floatleft text_result">
				<span><?php echo plugin_lang_get( 'dialog_text' ) ?></span>
			</div>
			<div class="action_div floatright">
				<input type="submit" class="button_submit" value="<?php echo plugin_lang_get( 'button_update' ) ?>" />
			</div>
		</div>
		<fieldset>
			<?php echo form_security_field( 'ajax_form', $f_token ) ?>
			<input type="hidden" name="target" value="<?php echo $p_target ?>" />
			<input type="hidden" name="project_id" value="<?php echo $t_project_id ?>" />
			<ol class="sortable">
				<?php foreach( $t_view_columns as $t_column ) { ?>
				<li>
					<input type="checkbox" name="columns[]" value="<?php echo $t_column ?>" id="<?php echo $p_target . '_' . $t_column ?>" checked="checked" />
					<label for="<?php echo $p_target . '_' . $t_column ?>"><?php echo $t_column ?></label>
				</li>
				<?php } ?>
				<?php foreach( $t_deselected_columns as $t_column ) { ?>
				<li>
					<input type="checkbox" name="columns[]" value="<?php echo $t_column ?>" id="<?php echo $p_target . '_' . $t_column ?>" />
					<label for="<?php echo $p_target . '_' . $t_column ?>"><?php echo $t_column ?></label>
				</li>
				<?php } ?>
			</ol>
		</fieldset>
	</form>
	<?php
}


if( !@form_security_validate( 'ajax_form' ) ) {
	exit;
}
$f_project_id = gpc_get( 'project_id', null );
if( null === $f_project_id ) {
	header( ' ', true, 400 );
	exit;
}
$g_project_override = $f_project_id;

?>
<div class="jqui_tabs">
	<ul class="tabs">
		<li><a href='#tab1'><?php echo lang_get( 'view_issues_columns_title' ) ?></a></li>
		<li><a href='#tab2'><?php echo lang_get( 'print_issues_columns_title' ) ?></a></li>
		<li><a href='#tab3'><?php echo lang_get( 'csv_columns_title' ) ?></a></li>
		<li><a href='#tab4'><?php echo lang_get( 'excel_columns_title' ) ?></a></li>
	</ul>
	<div class="columns_form_inner">

		<div id="tab1" class="columns_form_container">
			<?php print_columns_inputs_by_target( COLUMNS_TARGET_VIEW_PAGE ) ?>
		</div>

		<div id="tab2" class="columns_form_container">
			<?php print_columns_inputs_by_target( COLUMNS_TARGET_PRINT_PAGE ) ?>
		</div>

		<div id="tab3" class="columns_form_container">
			<?php print_columns_inputs_by_target( COLUMNS_TARGET_CSV_PAGE ) ?>
		</div>

		<div id="tab4" class="columns_form_container">
			<?php print_columns_inputs_by_target( COLUMNS_TARGET_EXCEL_PAGE ) ?>
		</div>

	</div>
</div>
<?php






InlineColumnConfiguration/pages/ajax_form_update.php
<?php

$t_option = null;
$ok = true;
$error_msg = '';

if( current_user_is_protected() ) {
	$ok = false;
	$error_msg = 'User is protected, not allowed to change preferences';
}

if( $ok && !@form_security_validate( 'ajax_form' ) ) {
	$ok = false;
	$error_msg = 'Form has expired, reload';
}

$f_project_id = @gpc_get( 'project_id', null );
$f_target = @gpc_get( 'target', null );
$f_columns = @gpc_get_string_array( 'columns', array() );


if( $ok && ( null === $f_project_id || null === $f_target ) ) {
		$ok = false;
	$error_msg = 'Missing parameters';
}

switch( $f_target ) {
	case COLUMNS_TARGET_VIEW_PAGE:
		$t_option = 'view_issues_page_columns';
		break;
	case COLUMNS_TARGET_CSV_PAGE:
		$t_option = 'csv_columns';
		break;
	case COLUMNS_TARGET_PRINT_PAGE:
		$t_option = 'print_issues_page_columns';
		break;
	case COLUMNS_TARGET_EXCEL_PAGE:
		$t_option = 'excel_columns';
		break;
	default:
		$ok = false;
		$error_msg = 'Column type is not valid';
		break;
}



if( $ok ) {
	$t_all_columns = columns_get_all();
	$ok = @columns_ensure_valid( $t_option, $f_columns, $t_all_columns );
	if( !$ok ) {
		$error_msg = 'There is an invalid column name';
	}
}

if( $ok ) {
	$t_user_id = auth_get_current_user_id();
	if( json_encode( config_get( $t_option, '', $t_user_id, $f_project_id ) ) !== json_encode( $t_option ) ) {
		$ok = @config_set( $t_option, $f_columns, $t_user_id, $f_project_id );
	}
	if( !$ok ) {
		$error_msg = 'Error saving configuration';
	}
}


if( $ok ) {
	echo '<span class="result_ok">';
	echo plugin_lang_get( 'result_text_ok' );
	if( COLUMNS_TARGET_VIEW_PAGE == $f_target ){
		echo '<br />' . plugin_lang_get( 'result_text_reload' );
	}
	echo '</span>';
}
else {
	echo '<span class="result_error">ERROR: '. $error_msg . '</span>';
}



InlineColumnConfiguration/files/inline-column-configuration.js
jQuery(document).ready(function($) {
	
	var InlineColumnConfiguration = {
		copyValuesToHiddenField : function(target, fieldName) {

			var columns = Array();
			
			$("#columns_" + target + " :checked").each(function(index, element) {
				columns.push(element.id.substring(2));
			});
			
			$('[name=' + fieldName + ']').val(columns.join());
		}	
		
	}
	
	$('div').on('click', 'a[href*="#inline-column-configuration"]', function(e) {
		e.preventDefault();
		$('#manage-view-forms').attr({"class":"nohidden"});
		$('#manage-view-forms').dialog({ width: 'auto' });
		$('.sortable').sortable().disableSelection();
	});
	
	$('#manage-columns-form').submit(function(e) {
		InlineColumnConfiguration.copyValuesToHiddenField('1', 'view_issues_columns');
		InlineColumnConfiguration.copyValuesToHiddenField('2', 'print_issues_columns');
		InlineColumnConfiguration.copyValuesToHiddenField('3', 'csv_columns');
		InlineColumnConfiguration.copyValuesToHiddenField('4', 'excel_columns');
	});
});


InlineColumnConfiguration.php
<?php
# Copyright (c) 2011 Robert Munteanu (robert@lmn.ro)

# Inline column configuration for MantisBT is free software: 
# you can redistribute it and/or modify it under the terms of the GNU
# General Public License as published by the Free Software Foundation, 
# either version 2 of the License, or (at your option) any later version.
#
# Inline column configuration plugin for MantisBT is distributed in the hope 
# that it will be useful, but WITHOUT ANY WARRANTY; without even the 
# implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  
# See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Inline column configuration plugin for MantisBT.  
# If not, see <http://www.gnu.org/licenses/>.

class InlineColumnConfigurationPlugin extends MantisPlugin {

	public function register() {
		$this->name = plugin_lang_get("title");
		$this->description = plugin_lang_get("description");

		$this->version = "2.0.0";

		$this->requires = array(
			"MantisCore" => "2.2",
			"jQueryUI" => "1.8"
		);

		$this->author = "Robert Munteanu, Carlos Proensa";
		$this->contact = "robert@lmn.ro";
		$this->url ="http://www.mantisbt.org/wiki/doku.php/mantisbt:inlinecolumnconfiguration";
		
		$this->scripts = array(
			'view_all_bug_page.php',
		);
	}

	#Check for conditions when this plugin is allowed to hook
	function check_page() {
		return auth_is_user_authenticated() && !current_user_is_protected() &&  in_array( basename( $_SERVER['SCRIPT_NAME'] ), $this->scripts );
	}	
	
	public function hooks() {
		$h = array(
			'EVENT_MENU_FILTER' => 'add_configure_columns_link',
			'EVENT_LAYOUT_RESOURCES' => 'resources',
			'EVENT_LAYOUT_BODY_END' => 'add_columns_dialog'
		);
		return $h;
	}

	public function resources( $p_event ) {
		if( $this->check_page() ) {
			return '<script type="text/javascript" src="' . plugin_file( 'inline-column-configuration.js' ) . '"></script>'
				 . '<link rel="stylesheet" type="text/css" href="'. plugin_file( 'inline-column-configuration.css' ) .'"/>';
		}
	}

	public function add_configure_columns_link() {
		if( !$this->check_page() ) {
			return;
		}
		$t_token = form_security_token( 'ajax_form' );
		$t_url = plugin_page( 'ajax_form' ) . '&project_id=' . helper_get_current_project() . '&ajax_form_token=' . $t_token;
		$t_link = '<a href="account_manage_columns_page.php" data-remote="' . $t_url . '" class="columns_form_trigger btn btn-sm btn-primary btn-white btn-round">'
				. plugin_lang_get( 'configure_columns' )
				. '</a>';
		return $t_link;
	}

	public function add_columns_dialog() {
		if( !$this->check_page() ) {
			return;
		}
		$t_title = lang_get( 'manage_columns_config' ) . ' (' . lang_get( 'email_project' ) . ': ' . project_get_name( helper_get_current_project() ) . ')';
		?>
		<div id="column_config_dialog" class="dialog" title="<?php echo $t_title ?>">
		<?php echo plugin_lang_get( 'please_wait' ) ?>
		</div>
		<?php
	}
	
}
